transaction/std/src/memo_builder/gift_code_memo_builder.rs

```diff
@@ -0,0 +1,181 @@
+
+use super::{memo::UnusedMemo, MemoBuilder};
+use crate::{
+    memo::{GiftCodeCancellationMemo, GiftCodeFundingMemo, GiftCodeSenderMemo},
+    ChangeDestination,
+};
+use mc_account_keys::PublicAddress;
+use mc_crypto_keys::RistrettoPublic;
+use mc_transaction_core::{Amount, MemoContext, MemoError, MemoPayload, NewMemoError};
+use std::convert::TryFrom;
+
+/// The three possible memo types that can be written as specified in MCIP #32
+/// | Memo type bytes | Name                                              |
+/// | -----------     | -----------                                       |
+/// | 0x0002          | Gift Code Sender Memo                             |
+/// | 0x0201          | Gift Code Funding Memo                            |
+/// | 0x0202          | Gift Code Cancellation Memo                       |
+#[derive(Clone, Debug)]
+pub enum GiftCodeMemoType {
+    /// Gift Code Sender Memo 0x0002
+    Sender,
+    /// Gift Code Funding Memo 0x0201
+    Funding,
+    /// Gift Code Cancellation Memo 0x0202
+    Cancellation,
+}
+
+/// This memo builder writes gift code memos to the change address which can be
+/// used by apps that offer gift codes as feature to track the status of gift
+/// codes as well as certain contextual information about them. This builder
+/// writes 3 possible gift code memo types. Only one memo which can be specified
+/// at the creation of the memo builder. It is worth noting that all memos
+/// written by this memo builder will go to the reserved change address and
+/// does not write any destination memos. For normal recoverable transaction
+/// history (RTH), the RTH memo builder should be used in conjuction with this
+/// memo builder.
+#[derive(Clone, Debug)]
+pub struct GiftCodeMemoBuilder {
+    // The gift code memo type we'll be writing
+    gift_code_memo_type: GiftCodeMemoType,
+    // The utf-8 formatted note that appears on the change address when a gift code is
+    // or cancelled
+    note: &'static str,
+    // TxOut Public Key of the gift code if it it's a funding memo
+    gift_code_tx_out_public_key: Option<RistrettoPublic>,
+    // Global TxOut index of the gift code if it's a cancellation memo
+    gift_code_tx_out_global_index: Option<u64>,
+    // Whether or not to enable gift code change memos
+    gift_code_change_memo_enabled: bool,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+}
+
+impl GiftCodeMemoBuilder {
+    /// Create a new gift code memo builder specifying the memo type
+    pub fn new(gift_code_memo_type: GiftCodeMemoType) -> Self {
+        Self {
+            gift_code_memo_type,
+            note: "",
+            gift_code_tx_out_public_key: None,
+            gift_code_tx_out_global_index: None,
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+        }
+    }
+    /// A note in the form of null terminated utf-8 bytes can be applied to
```

Comment:
It may just be my unfamiliarity with rust but the "null terminated utf-8 bytes" made me do a double take as I looked at the input type of `str`.  Is this not captured in the definition of `str`?  It leaves one wondering if `60` bytes is meant to include the null or not.  Looking at the funding and sender memos It looks like it doesn't include the null terminator

Reply:
There's discussion on the Gift Codes PR about making it "null terminated unless max length"